src/relay/pass/type_solver.cc

```diff
@@ -459,51 +647,93 @@ Type TypeSolver::Resolve(const Type& type) {
-      CHECK_LE(args.size(), rel->args.size());
-    }
-
-    CHECK(rnode->location.defined())
-      << "undefined location, should be set when constructing relation node";
-
-    // We need to set this in order to understand where unification
-    // errors generated by the error reporting are coming from.
-    reporter_->SetLocation(rnode->location);
-
-    try {
-      // Call the Type Relation's function.
-      bool resolved = rel->func(args, rel->num_inputs, rel->attrs, reporter_);
-
-      if (resolved) {
-        ++num_resolved_rels_;
+  while (true) {
+    while (!update_queue_.empty()) {
+      RelationNode* rnode = update_queue_.front();
+      const auto& rel = rnode->rel;
+      update_queue_.pop();
+      CHECK(!rnode->resolved);
+      // update the relation with given evidence.
+      Array<Type> args;
+      for (auto* tlink = rnode->type_list.head; tlink != nullptr; tlink = tlink->next) {
+        args.push_back(Resolve(tlink->value->FindRoot()->resolved_type));
+        CHECK_LE(args.size(), rel->args.size());
+      }
+
+      CHECK(rnode->location.defined())
+        << "undefined location, should be set when constructing relation node";
+
+      // We need to set this in order to understand where unification
+      // errors generated by the error reporting are coming from.
+      reporter_->SetLocation(rnode->location);
+
+      try {
+        // Call the Type Relation's function.
+        bool resolved = rel->func(args, rel->num_inputs, rel->attrs, reporter_);
+
+        if (resolved) {
+          ++num_resolved_rels_;
+        }
+
+        rnode->resolved = resolved;
+      } catch (const Error& err) {
+        this->ReportError(err, rnode->location);
+        rnode->resolved = false;
+      } catch (const dmlc::Error& err) {
+        rnode->resolved = false;
+        this->ReportError(
+            RELAY_ERROR(
+              "an internal invariant was violated while " \
+              "typechecking your program " <<
+              err.what()), rnode->location);
       }
 
-      rnode->resolved = resolved;
-    } catch (const Error& err) {
-      this->ReportError(err, rnode->location);
-      rnode->resolved = false;
-    } catch (const dmlc::Error& err) {
-      rnode->resolved = false;
-      this->ReportError(
-          RELAY_ERROR(
-            "an internal invariant was violated while " \
-            "typechecking your program " <<
-            err.what()), rnode->location);
-    }
-
-    // Mark inqueue as false after the function call
-    // so that rnode itself won't get enqueued again.
-    rnode->inqueue = false;
+      // Mark inqueue as false after the function call
+      // so that rnode itself won't get enqueued again.
+      rnode->inqueue = false;
+    }
+
+    bool solved = false;
+
+    for (auto type : type_nodes_) {
+      auto root = type->FindRoot();
+      auto rty = root->resolved_type;
+      if (rty.as<IncompleteTypeNode>() != nullptr && root->hint.defined()) {
+        solved = true;
```

Comment:
`solved` seems like a misnomer since all that is being checked is whether the hints provide new information for more checking, not whether everything has been solved